#' Compute and plot empirical codon usage indexes based on the inferred P-sites.
#'
#' For a specified sample this function computes an empirical codon usage index
#' based on the frequency of in-frame P-sites along the coding sequence. It
#' computes the codon usage index for all the 64 triplets, normalizes them for
#' the frequency of each codon within the CDS and returns a bar plot with the
#' resulting values. This function also allows to compare the computed codon
#' usage indexes with a set of 64 values provided by the user.
#'
#' @param data A list of data tables from \code{\link{psite_info}} that may or
#'   may not include a column named \emph{psite_codon}. This column reports the
#'   three nucleotides covered by the P-site and can be previously generated by
#'   the \code{\link{psite_info}} function. If not already present, it is
#'   automatically generated throughout the function starting from a FASTA file
#'   or a BSgenome data package.
#' @param annotation A data table as generated by
#'   \code{\link{create_annotation}}.
#' @param sample A character string vector specifying the name of the sample of
#'   interest.
#' @param fastapath An optional character string specifying the path to the
#'   FASTA file used in the alignment step, including its name and extension.
#'   This file can contain reference nucleotide sequences either of a genome
#'   assembly or of all the transcripts (see \code{fasta_genome}). Please make
#'   sure the sequences derive from the same release of the annotation file used
#'   in the \code{\link{create_annotation}} function. Note: either
#'   \code{fastapath} or \code{bsgenome} is required to normalize the data, even
#'   if a column \emph{psite_codon} has been previously generated by
#'   \code{\link{psite_info}}. Default is NULL.
#' @param fasta_genome A locigal value whether or not the FASTA file specified
#'   by \code{fastapath} contains nucleotide sequences of a genome assembly.
#'   FALSE means that the nucleotide sequences of all the transcripts are
#'   provided instead. When this parameter is TRUE (the default), an annotation
#'   object is required (see \code{gtfpath} and \code{txdb}).
#' @param bsgenome An optional character string specifying the name of the
#'   BSgenome data package containing the genome sequences to be loaded. If it
#'   is not already present in your system, it will be installed through the
#'   biocLite.R script (check the list of data packages available in the
#'   Bioconductor repositories for your version of R/Bioconductor by the
#'   \code{\link[BSgenome]{available.genomes}} function of the BSgenome
#'   package). This parameter also requires an annotation object (see
#'   \code{gtfpath} and \code{txdb}). Please make sure the sequences included in
#'   the specified BSgenome data pakage are in agreement with the sequences used
#'   in the alignment step. Note: either \code{fastapath} or \code{bsgenome} is
#'   required to normalize the data, even if a column \emph{psite_codon} has
#'   been previously generated by \code{\link{psite_info}}. Default is NULL.
#' @param gtfpath A character string specifying the path to te GTF file,
#'   including its name and extension. Please make sure the GTF derives from the
#'   same release of what is specified by \code{fastapath} or by
#'   \code{bsgenome}. Note that either \code{gtfpath} or \code{txdb} must be
#'   specified when the nucleotide sequences of a genome assembly are provided
#'   (see \code{fastapath} or \code{bsgenome}). Default is NULL.
#' @param txdb A character string specifying the name of the annotation package
#'   for TxDb object(s) to be loaded. If it is not already present in your
#'   system, it will be installed through the biocLite.R script (check the list
#'   of TxDb annotation packages available in the Bioconductor repositories at
#'   http://bioconductor.org/packages/release/BiocViews.html#___TxDb )). Please
#'   make sure the annotation package derives from the same release of what is
#'   specified by \code{fastapath} or by \code{bsgenome}. Note that either
#'   \code{gtfpath} or \code{txdb} must be specified when the nucleotide
#'   sequences of a genome assembly are provided (see \code{fastapath} or
#'   \code{bsgenome}). Default is NULL.
#' @param dataSource An optional character string describing the origin of the
#'   GTF data file. For more information about this parameter please refer to
#'   the description of \emph{dataSource} of the
#'   \code{\link[GenomicFeatures]{makeTxDbFromGFF}} function included in the
#'   \code{GenomicFeatures} package.
#' @param organism A optional character string reporting the genus and species
#'   of the organism when \code{gtfpath} is specified.  For more information
#'   about this parameter please refer to the description of \emph{dataSource}
#'   of the \code{\link[GenomicFeatures]{makeTxDbFromGFF}} function included in
#'   the \code{GenomicFeatures} package.
#' @param transcripts A character string vector specifying the name of the
#'   transcripts to be included in the analysis. By default this argument is
#'   NULL, meaning that all the transcripts in \code{data} are used. Please note
#'   that the transcripts not associated to any annotated \emph{5' UTR},
#'   \emph{CDS} and \emph{3'UTR} and transcripts with coding sequence length not
#'   divisible by 3 are automatically discarded.
#' @param codon_values A data table containing codon-specific values provided by
#'   the user. These values are compared with the empirical codon usage indexes
#'   of the sample of interest. The data table must contain at least the 64
#'   codons and the corresponding values arranged in two columns named
#'   \emph{codon} and \emph{value}, respectively. Note that a similar data
#'   table is also returned by \code{\link{codon_usage_psite}} itself. Default
#'   is NULL.
#' @param scatter_label A logical value whether or not to label the dots of the
#'   scatter plot generated by specifying \code{codon_values}. Each dot can be
#'   labeled either after the three nucleotides of the codon or after the
#'   corresponding amino acid (see \code{aminoacid}). This parameter is
#'   considered only if \code{codon_values} is specified. Default is FALSE.
#' @param aminoacid A logical value whether or not to label the dots of the
#'   scatter plot generated by specifying \code{codon_values} using the amino
#'   acids corresponding to the triplets. Default is FALSE, meaning that the
#'   three nucleotides of the codon are used instead. This parameter is
#'   considered only if \code{codon_values} is specified and \code{scatter_label}
#'   is TRUE. Default is FALSE.
#' @return A list containing a ggplot2 object (named "plot"), and a data table
#'   ("dt") with the associated data. An additional ggplot2 object
#'   ("plot_comparison") is returned if \code{codon_values} is specified.
#' @import data.table
#' @import ggplot2
#' @export
codon_usage_psite <- function(data, annotation, sample, fastapath = NULL, fasta_genome = TRUE,
                              bsgenome = NULL, gtfpath = NULL, txdb = NULL, 
                              dataSource = NA, organism = NA,
                              transcripts = NULL, codon_values = NULL,
                              scatter_label = FALSE, aminoacid = FALSE) {
  
  if(length(codon_values) != 0){
    rownames(codon_values) <- as.character(codon_values$codon)
    if (nrow(codon_values) < 64 | length(unique(rownames(codon_values))) < 64){
      cat("\n")
      stop("number of different triplets in codon_values < 64. At least one codon is missing\n\n")
    } else {
      if (length(unique(rownames(codon_values))) > 64){
        cat("\n")
        stop("number of different triplets in codon_values > 64. Too many codons\n\n")
      }
    }
  }
  
  if(length(fastapath) == 0 & length(bsgenome) == 0){
    cat("\n")
    stop("impossible to normalize the data: no nucleotide sequences are provided. Either fastpath or bsgenome must be specified\n\n")
  }

  if(((length(fastapath) != 0 & (fasta_genome == TRUE | fasta_genome == T)) |
      length(bsgenome) != 0) &
     length(gtfpath) == 0 & length(txdb) == 0){
    cat("\n")
    stop("genome annotation file not specified (both GTF path and TxDb object are missing)\n\n")
  }
  
  if(length(fastapath) != 0 & length(bsgenome) != 0){
    cat("\n")
    warning("both fastapath and bsgenome are specified. Only fastapath will be considered\n")
    bsgenome = NULL
  }
  
  if(length(gtfpath) != 0 & length(txdb) != 0){
    cat("\n")
    warning("both gtfpath and txdb are specified. Only gtfpath will be considered\n")
    txdb = NULL
  }
  
  if((length(gtfpath) != 0 | length(txdb) != 0) &
     ((length(fastapath) == 0 & length(bsgenome) == 0) |
      (length(fastapath) != 0 & (fasta_genome == FALSE | fasta_genome == F)))){
    cat("\n")
    warning("a genome annotation file is specified but no sequences from genome assembly are provided\n")
  }
  
  if(length(gtfpath) != 0 | length(txdb) != 0){
    if(length(gtfpath) != 0){
      path_to_gtf <- gtfpath
      txdbanno <- GenomicFeatures::makeTxDbFromGFF(file=path_to_gtf, format="gtf", dataSource = dataSource, organism = organism)
    } else {
      if(txdb %in% rownames(installed.packages())){
        library(txdb, character.only = TRUE)
      } else {
        source("https://bioconductor.org/biocLite.R")
        biocLite(txdb, suppressUpdates = TRUE)
        library(txdb, character.only = TRUE)
      }
      txdbanno <- get(txdb)
    }
  }
  
  if(length(fastapath) != 0 | length(bsgenome) != 0){
    if(length(fastapath) != 0) {
      if(fasta_genome == TRUE | fasta_genome == T){
        temp_sequences <- Biostrings::readDNAStringSet(fastapath, format = "fasta", use.names = TRUE)
        exon <- suppressWarnings(GenomicFeatures::exonsBy(txdbanno, by="tx", use.names=TRUE))
        exon <- as.data.table(exon[unique(names(exon))])
        sub_exon <- exon[seqnames %in% names(temp_sequences)]
        seq_df <- sub_exon[, list(seq = paste(Biostrings::subseq(temp_sequences[as.character(seqnames)],
                                                                 start = start,
                                                                 end = end), collapse="")),
                           by = group_name]
        sequences <- Biostrings::DNAStringSet(seq_df$seq)
      } else {
        sequences <- Biostrings::readDNAStringSet(fastapath, format = "fasta", use.names = TRUE)
      }
    } else {
      if(bsgenome %in% installed.genomes()){
        library(bsgenome, character.only = TRUE)
      } else {
        source("http://www.bioconductor.org/biocLite.R")
        biocLite(bsgenome, suppressUpdates = TRUE)
        library(bsgenome, character.only = TRUE)
      }
      sequences <- GenomicFeatures::extractTranscriptSeqs(get(bsgenome), txdbanno, use.names=T)
    }
  }
  
  if(!("psite_codon" %in% colnames(data[[sample]]))){
    data[[sample]][, psite_codon := as.character(Biostrings::subseq(sequences[as.character(data[[sample]]$transcript)],
                                                                    start = data[[sample]]$psite,
                                                                    end = data[[sample]]$psite + 2))]
  }
  
  l_transcripts <- as.character(annotation[l_utr5 > 0 & 
                                             l_cds >0 &
                                             l_cds %% 3 == 0 &
                                             l_utr3 > 0, transcript])
  
  cod_aa <- data.table(codon=c("GCC", "GCG", "GCU", "GCA", "AGA", "CGG", "AGG", "CGA", "CGC", "CGU", "AAC", "AAU", "GAC", "GAU", "UGC", "UGU", "CAA", "CAG", "GAG", "GAA", "GGC", "GGU", "GGA", "GGG", "CAC", "CAU", "AUA", "AUC", "AUU", "CUG", "CUA", "UUA", "CUU", "UUG", "CUC", "AAA", "AAG", "AUG", "UUC", "UUU", "CCG", "CCC", "CCU", "CCA", "AGC", "UCG", "UCU", "UCA", "UCC", "AGU", "UAG", "UAA", "UGA", "ACA", "ACC", "ACG", "ACU", "UGG", "UAU", "UAC", "GUA", "GUG", "GUU", "GUC"),
                       aa=c("A", "A", "A", "A", "R", "R", "R", "R", "R", "R", "N", "N", "D", "D", "C", "C", "Q", "Q", "E", "E", "G", "G", "G", "G", "H", "H", "I", "I", "I", "L", "L", "L", "L", "L", "L", "K", "K", "M", "F", "F", "P", "P", "P", "P", "S", "S", "S", "S", "S", "S", "*", "*", "*", "T", "T", "T", "T", "W", "Y", "Y", "V", "V", "V", "V"))
  cod_lev <- gsub("U", "T", cod_aa$codon)
  
  if (length(transcripts) == 0) {
    c_transcript <- l_transcripts
  } else {
    c_transcript <- intersect(l_transcripts, transcripts)
  }

  sub_sequences <- sequences[c_transcript]
  cds_biost <- Biostrings::subseq(sub_sequences,
                      start = annotation[transcript %in% names(sub_sequences), l_utr5] + 1,
                      end = annotation[transcript %in% names(sub_sequences), l_utr5] +
                        annotation[transcript %in% names(sub_sequences), l_cds])
  
  seq_freq <- (Biostrings::trinucleotideFrequency(cds_biost, step = 3,
                                                  as.prob = TRUE,
                                                  with.labels = TRUE,
                                                  simplify.as = "collapsed")) * 1000
  names(seq_freq) <- gsub("T", "U", names(seq_freq))
  
  dt <- data[[sample]][as.character(transcript) %in% c_transcript & psite_region == "cds" & psite_from_start %% 3 == 0]
  norm_table <- data.table(table(factor(dt$psite_codon, levels = cod_lev)))
  colnames(norm_table) <- c("codon", "count")
  norm_table[, codon := gsub("T", "U", codon)]
  norm_table[, count := (count / sum(count)) * 1000]

  norm_table <- norm_table[, class := "cds"
                           ][codon == "AUG", class := "start"
                             ][codon %in% c("UAA", "UGA", "UAG"), class := "stop"
                               ][, class := factor(class, levels = c("start", "cds", "stop"), labels=c("Start codon ", "cds", "Stop codon"))
                                 ][cod_aa, on = "codon"
                                   ][, value := count / seq_freq[codon]
                                     ][, value := value - min(value)
                                       ][, value := value / max(value)
                                         ][order(value)
                                           ][, codon := factor(codon, levels = codon)]
  
  colour_codon <- ifelse(norm_table$codon == "AUG", "#104ec1",
                         ifelse(norm_table$codon %in% c("UAA", "UGA", "UAG"),
                                "darkred", "gray40"))

  bs <- 30
  bp <- ggplot(norm_table,aes(x = codon, y = value, fill = class)) +
    geom_bar(stat = "identity", alpha = 0.9) +
    scale_fill_manual(name = "", breaks=c("Start codon ","Stop codon"), values = c("#104ec1", "gray60", "darkred")) +
    theme_bw(base_size = bs) +
    theme(legend.position = "top") +
    scale_x_discrete("Codon") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5, size = bs * 0.5)) +
    geom_text(aes(label = aa), vjust = -0.5) +
    theme(axis.text.x = element_text(colour = colour_codon)) +
    scale_y_continuous("Usage index", limits = c(0,1.05), breaks = c(0,0.5,1))
  
  output <- list()
  output[["plot"]] <- bp
  output[["dt"]] <- norm_table[, c("codon", "class", "value",  "aa")]

  if(length(codon_values) != 0){
    colnames(codon_values) <- c("codon", "comp_values")
    codon_values[, codon := gsub("T", "U", codon)]
    norm_table <- norm_table[comp_values, on = "codon"
                             ][, comp_values := comp_values - min(comp_values)
                               ][, comp_values := comp_values / max(comp_values)]
               
    correlation <- round(cor(norm_table$value, norm_table$comp_values), 3)
    
    bs <- 30
    pcomp <- ggplot(norm_table,aes(x = value, y = comp_values, colour = class)) +
      geom_smooth(method = "lm", se=T, color="gray80", fill="gray80", linetype = 1, formula = y ~ x, level = 0.99,  fullrange = TRUE) +
      geom_point(alpha = 0.9, size = bs * 0.9) +
      ggrepel::geom_text_repel(aes(value, comp_values, label = codon), show.legend = F) +
      scale_colour_manual(name = "", breaks = c("Start codon ","Stop codon"), values = c("#104ec1", "gray40", "darkred")) +
      theme_bw(base_size = bs) +
      theme(legend.position = "top") +
      scale_x_continuous("Codon usage index", limits = c(-0.3,1.3), breaks = c(0,0.25,0.5,0.75,1), expand = c(0,0)) +
      scale_y_continuous("Codon usage", limits = c(-0.3,1.3), breaks = c(0,0.25,0.5,0.75,1), expand = c(0,0)) +
      coord_cartesian(xlim = c(-0.05,1.05), ylim = c(-0.05,1.05)) +
      geom_text(aes(x = 1, y = 0), label = paste0("R=",correlation), vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black")
    pcomp
    
    if (scatter_label == T || scatter_label == TRUE) {
      if (aminoacid == T || aminoacid == TRUE) {
        pcomp <- pcomp +
          ggrepel::geom_text_repel(aes(value, comp_values, label = as.character(aa)), show.legend = F)
      } else {
        pcomp <- pcomp +
          ggrepel::geom_text_repel(aes(value, comp_values, label = as.character(codon)), show.legend = F)
      }
    }
    
    output[["plot_comparison"]] <- pcomp
  }
  return(output)
}
