#' Plot metaheatmaps based on the two extremities of the reads.
#'
#' For a specified sample this function plots four metaheatmaps showing the
#' abundance of the 5' and the 3' end of the reads mapping around the start and
#' the stop codon of the annotated CDS (if any), stratified by their length. It
#' is possible to visualise the metaheatmaps for all the read lengths or to
#' restrict the graphical output to a sub-range of read lengths.
#'
#' @param data A list of data tables from either \code{\link{bamtolist}} or
#'   \code{\link{bedtolist}}.
#' @param annotation A data table as generated by
#'   \code{\link{create_annotation}}.
#' @param sample A character string specifying the name of the sample of
#'   interest.
#' @param transcripts A character string vector specifying the name of the
#'   transcripts to be included in the analysis. By default this argument is
#'   NULL, meaning that all the transcripts in \code{data} are used. Note that
#'   if either the 5' UTR, the coding sequence or the 3' UTR of a transcript is
#'   shorther than \code{utr5l}, \eqn{2*}\code{cdsl} and \code{utr3l}
#'   respectively, the transcript is automatically discarded.
#' @param cl An integer value in \emph{[1,100]} specifying the confidence level
#'   for restricting the plot to a sub-range of read lengths. By default it is
#'   set to 99.
#' @param utr5l A positive integer specifying the length (in nucleotides) of the
#'   5' UTR region that in the plot flanks the start codon. The default value is
#'   50.
#' @param cdsl A positive integer specifying the length (in nucleotides) of the
#'   CDS region that in the plot will flank both the start and the stop codon.
#'   The default value is 50.
#' @param utr3l A positive integer specifying the length (in nucleotides) of the
#'   3' UTR region that in the plot flanks the start codon. The default value is
#'   50.
#' @param log A logical value whether or not to use a logarithmic scale colour
#'   (strongly suggested in case of large variations of the signal). Default is
#'   FALSE.
#' @param colour A character string specifying the colour of the plot. Default
#'   is "black".
#' @return A list containing a ggplot2 object, and a data table with the
#'   associated data.
#' @examples
#' data(reads_list)
#' data(mm81cdna)
#'
#' ## Visualise the metaheatmaps for all the read lengths
#' heatend_whole <- rends_heat(reads_list, mm81cdna, sample = "Samp1", cl = 100)
#' heatend_whole[["plot"]]
#'
#' ## Visualise the metaheatmaps for a sub-range of read lengths (the middle
#' 95%) reducing the flanking regions around the start and the stop codon
#' heatend_sub95 <- rends_heat(reads_list, mm81cdna, sample = "Samp1", cl = 95,
#' utr5l = 30, cdsl = 40, utr3l = 30)
#' heatend_sub95[["plot"]]
#' @import data.table
#' @import ggplot2
#' @export
rends_heat <- function(data, annotation, sample, transcripts = NULL, cl = 99, utr5l = 50, cdsl = 50, utr3l = 50,
                       log = F, colour = "black") {
  temp_dt <- data[[sample]]
  temp_dt[, start_dist_end5 := end5 - start_pos
          ][, stop_dist_end5 := end5 - stop_pos
            ][, start_dist_end3 := end3 - start_pos
              ][, stop_dist_end3 := end3 - stop_pos]

  minlen <- ceiling(quantile(temp_dt$length, (1 - cl/100)/2))
  maxlen <- ceiling(quantile(temp_dt$length, 1 - (1 - cl/100)/2))
  
  l_transcripts <- as.character(annotation[l_utr5 >= utr5l & 
                                             l_cds > 2 * (cdsl + 1) &
                                             l_utr3 >= utr3l, transcript])
  
  if (length(transcripts) == 0) {
    c_transcripts <- l_transcripts
  } else {
    c_transcripts <- intersect(l_transcripts, transcripts)
  }
  
  dt <- temp_dt[transcript %in% c_transcripts]
  temp_dt[, c("start_dist_end5", "stop_dist_end5", "start_dist_end3", "stop_dist_end3") := NULL]
  
  # 5' end
  start_sub <- dt[start_dist_end5 %in% seq(-utr5l, cdsl)]
  start_tab <- setkey(start_sub, length, start_dist_end5)[CJ(unique(length), unique(start_dist_end5)), .N, by=.EACHI]
  setnames(start_tab, c("length", "dist", "count"))
  start_tab[, region := "start"]
  stop_sub <- dt[stop_dist_end5 %in% seq(-cdsl, utr3l)]
  stop_tab <- setkey(stop_sub, length, stop_dist_end5)[CJ(unique(length), unique(stop_dist_end5)), .N, by=.EACHI]
  setnames(stop_tab, c("length", "dist", "count"))
  stop_tab[, region := "stop"]
  final_tab5 <- rbind(start_tab, stop_tab)
  final_tab5[, end := "5end"]
  
  # 3' end
  start_sub <- dt[start_dist_end3 %in% seq(-utr5l, cdsl)]
  start_tab <- setkey(start_sub, length, start_dist_end3)[CJ(unique(length), unique(start_dist_end3)), .N, by=.EACHI]
  setnames(start_tab, c("length", "dist", "count"))
  start_tab[, region := "start"]
  stop_sub <- dt[stop_dist_end3 %in% seq(-cdsl, utr3l)]
  stop_tab <- setkey(stop_sub, length, stop_dist_end3)[CJ(unique(length), unique(stop_dist_end3)), .N, by=.EACHI]
  setnames(stop_tab, c("length", "dist", "count"))
  stop_tab[, region := "stop"]
  final_tab3 <- rbind(start_tab, stop_tab)
  final_tab3[, end := "3end"]

  # all
  final_tab <- rbind(final_tab5, final_tab3)
  final_tab[, region := factor(region, levels = c("start", "stop"), labels = c("Distance from start (nt)", "Distance from stop (nt)"))]
  final_tab[, end := factor(end, levels = c("5end", "3end"), labels = c("5' end", "3' end"))]

  # plot
  max<-max(final_tab$count)
  p <- ggplot(final_tab, aes(dist, length)) +
    geom_tile(aes(fill = count)) +
    labs(title = paste(sample, "5' / 3' read end metaheatmaps", sep = " - "), y = "Read length") +
    theme_bw(base_size = 20) +
    theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(), panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank(), axis.title.x = element_blank()) +
    facet_grid(end ~ region, scales = "free", switch = "x") +
    theme(strip.background = element_blank(), strip.placement = "outside") +
    theme(plot.title = element_text(hjust = 0.5)) +
    scale_y_continuous(limits = c(minlen - 0.5, maxlen + 0.5), breaks = seq(minlen + ((minlen) %% 2), maxlen, by = max(2,floor((maxlen - minlen)/7)))) +
    geom_vline(xintercept = 0, linetype = 2, color = "red")
  
  if (log == F) {
    p <- p +
      scale_fill_gradient("Number\nof read\nextremities\n", low = "white", high = colour, limits = c(0.1, max), breaks = c(0.1, max/2, max), labels = c("0", floor(max/2), floor(max)), na.value = "white")
  } else {
    p <- p +
      scale_fill_gradient("Number\nof read\nextremities\n", low = "white", high = colour, limits = c(0.1, max), breaks = c(0.1, 10^(log10(max)/2 - 0.5), floor(max)), labels = c("0", floor(10^(log10(max)/2 - 0.5)), floor(max)), trans = "log", na.value = "transparent")
  }
  
  output<-list()
  output[["plot"]]<-p
  output[["dt"]]<-final_tab
  return(output)
}
